src/hp/magicsock/conn.rs

```diff
@@ -0,0 +1,3379 @@
+        r.await.unwrap();
+    }
+}
+
+/// The info and state for the DiscoKey in the Conn.discoInfo map key.
+///
+/// Note that a DiscoKey does not necessarily map to exactly one
+/// node. In the case of shared nodes and users switching accounts, two
+/// nodes in the NetMap may legitimately have the same DiscoKey.  As
+/// such, no fields in here should be considered node-specific.
+pub(super) struct DiscoInfo {
+    pub(super) node_key: key::node::PublicKey,
+    /// The precomputed key for communication with the peer that has the `node_key` used to
+    /// look up this `DiscoInfo` in Conn.discoInfo.
+    /// Not modified once initialized.
+    shared_key: key::node::SharedSecret,
+
+    /// Tthe src of a ping for `node_key`.
+    last_ping_from: Option<SocketAddr>,
+
+    /// The last time of a ping for `node_key`.
+    last_ping_time: Option<Instant>,
+}
+
+/// Reports whether x and y represent the same set of endpoints. The order doesn't matter.
+fn endpoint_sets_equal(xs: &[cfg::Endpoint], ys: &[cfg::Endpoint]) -> bool {
+    if xs.is_empty() && ys.is_empty() {
+        return true;
+    }
+    if xs.len() == ys.len() {
+        let mut order_matches = true;
+        for (i, x) in xs.iter().enumerate() {
+            if x != &ys[i] {
+                order_matches = false;
+                break;
+            }
+        }
+        if order_matches {
+            return true;
+        }
+    }
+    let mut m: HashMap<&cfg::Endpoint, usize> = HashMap::new();
+    for x in xs {
+        *m.entry(x).or_default() |= 1;
+    }
+    for y in ys {
+        *m.entry(y).or_default() |= 2;
+    }
+
+    m.values().all(|v| *v == 3)
+}
+
+impl AsyncUdpSocket for Conn {
+    #[instrument(skip_all, fields(self.name = %self.name))]
+    fn poll_send(
+        &self,
+        _udp_state: &quinn_udp::UdpState,
+        cx: &mut Context,
+        transmits: &[quinn_udp::Transmit],
+    ) -> Poll<io::Result<usize>> {
+        let bytes_total: usize = transmits.iter().map(|t| t.contents.len()).sum();
+        record!(MagicsockMetrics::SendData, bytes_total as _);
+
+        if self.is_closed() {
+            record!(MagicsockMetrics::SendDataNetworkDown, bytes_total as _);
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "connection closed",
+            )));
+        }
+
+        let mut n = 0;
+        if transmits.is_empty() {
+            return Poll::Ready(Ok(n));
+        }
+
+        // Split up transmits by destination, as the rest of the code assumes single dest.
+        let groups = TransmitIter::new(transmits);
+        for group in groups {
+            match self.network_sender.try_reserve() {
+                Err(mpsc::error::TrySendError::Full(_)) => {
+                    // TODO: add counter?
+                    self.network_send_wakers
+                        .lock()
+                        .unwrap()
+                        .replace(cx.waker().clone());
```

Comment:
Probably does not make a big difference, but I think you only have to store the waker if you return pending.

If this returns with n>0 we will return Ready.